### PR TITLE
[WIP] feat: client generated by openapi-generator

### PR DESCRIPTION
**Pull request by tomplus**
The client has been generated by openapi-generator instead of swagger-codegen.

Ref: https://github.com/kubernetes-client/gen/issues/93

### CODE REVIEW

File: kubernetes/.openapi-generator/COMMIT [hunk 1]
```diff
@@ -0,0 +1,2 @@
+Requested Commit: v3.3.4
```
Comment (tomplus) [inline kubernetes/.openapi-generator/COMMIT]: `kubernete/.swagger-codegen` is deleted, openapi-generator uses `kubernetes/.openapi-generator` and also `.openapi-generator-ignore` instead of `.swagger-codegen-ignore` (renamed).

File: kubernetes/client/api/__init__.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
+from __future__ import absolute_import
+
+# flake8: noqa
+
+# import apis into api package
```
Comment (tomplus) [inline kubernetes/client/api/__init__.py]: This packages is renamed from apis to api.

File: setup.py [hunk 3]
```diff
@@ -56,7 +56,7 @@
     tests_require=TESTS_REQUIRES,
     extras_require=EXTRAS,
     packages=['kubernetes', 'kubernetes.client', 'kubernetes.config',
-              'kubernetes.watch', 'kubernetes.client.apis',
+              'kubernetes.watch', 'kubernetes.client.api',
```
Comment (tomplus) [inline setup.py]: `apis -> api`

File: kubernetes/e2e_test/test_batch.py [hunk 4]
```diff
@@ -16,7 +16,7 @@
 import uuid
 
 from kubernetes.client import api_client
-from kubernetes.client.apis import batch_v1_api
+from kubernetes.client.api import batch_v1_api
```
Comment (tomplus) [inline kubernetes/e2e_test/test_batch.py]: `apis -> api` here and in another e2e tests too.

File: scripts/rest_client_patch.diff [hunk 5]
```diff
@@ -1,8 +1,8 @@
 diff --git a/kubernetes/client/rest.py b/kubernetes/client/rest.py
-index 65fbe95..e174317 100644
+index 0bb15d8..acf7b2b 100644
 --- a/kubernetes/client/rest.py
 +++ b/kubernetes/client/rest.py
-@@ -152,6 +152,10 @@ class RESTClientObject(object):
+@@ -155,6 +155,10 @@ class RESTClientObject(object):
```
Comment (tomplus) [inline scripts/rest_client_patch.diff]: It needs to be updated to apply this patch by update_client.sh

File: kubernetes/client/__init__.py [hunk 6]
```diff
@@ -1,656 +1,657 @@
...
-    Generated by: https://github.com/swagger-api/swagger-codegen.git
+    OpenAPI spec version: v1.13.3
+    Generated by: https://openapi-generator.tech
 """
 
 
 from __future__ import absolute_import
 
-# import models into sdk package
-from .models.admissionregistration_v1beta1_service_reference import AdmissionregistrationV1beta1ServiceReference
```
Comment (tomplus) [inline kubernetes/client/__init__.py]: relative vs. absolute imports

File: kubernetes/client/api/admissionregistration_v1alpha1_api.py [hunk 7]
```diff
@@ -3,41 +3,39 @@
...
+    Generated by: https://openapi-generator.tech
 """
 
 
 from __future__ import absolute_import
 
-import sys
-import os
-import re
+import re  # noqa: F401
```
Comment (tomplus) [inline kubernetes/client/api/admissionregistration_v1alpha1_api.py]: There are a lot of changes which clean the code (unused imports, long lines etc.).

File: kubernetes/client/api/authentication_v1_api.py [hunk 8]
```diff
@@ -77,75 +76,74 @@ def create_token_review_with_http_info(self, body, **kwargs):
...
-            body_params = params['body']
+        if 'body' in local_var_params:
+            body_params = local_var_params['body']
         # HTTP header `Accept`
-        header_params['Accept'] = self.api_client.\
-            select_header_accept(['application/json', 'application/yaml', 'application/vnd.kubernetes.protobuf'])
-
-        # HTTP header `Content-Type`
-        header_params['Content-Type'] = self.api_client.\
-            select_header_content_type(['*/*'])
```
[the remark this answers is not in the document]
Comment (roycaihw) [reply kubernetes/client/api/authentication_v1_api.py]: I see. The `select_header_content_type` doesn't do what I expected https://github.com/kubernetes-client/python/issues/866#issuecomment-508510910

File: kubernetes/.openapi-generator/COMMIT [hunk 9]
```diff
@@ -0,0 +1,2 @@
...
+Actual Commit: 2353d71d4b02be6dbabe25aac1a9e56eb3b812a2
```
Comment (scottilee) [inline kubernetes/.openapi-generator/COMMIT]: Since its been a while can you bump this to `v4.0.3`, commit `203fc812bda4267020165e656f55185f72df4cc0`?

Comment (scottilee) [reply kubernetes/.openapi-generator/COMMIT]: Also just curious, how often would the version need to be bumped in the future?

File: kubernetes/README.md [hunk 10]
```diff
@@ -17,9 +17,9 @@ Python 2.7 and 3.4+
 If the python package is hosted on Github, you can install directly from Github
 
 ```sh
-pip install git+https://github.com/kubernetes-client/python.git
+pip install git+https://github.com/kubernetes-kubernetes.client/python.git
```
Comment (scottilee) [inline kubernetes/README.md]: Why are these changed to this weird URL?

File: kubernetes/client/__init__.py [hunk 11]
```diff
@@ -1,656 +1,657 @@
...
-from .models.v2beta2_metric_target import V2beta2MetricTarget
-from .models.v2beta2_metric_value_status import V2beta2MetricValueStatus
-from .models.v2beta2_object_metric_source import V2beta2ObjectMetricSource
-from .models.v2beta2_object_metric_status import V2beta2ObjectMetricStatus
-from .models.v2beta2_pods_metric_source import V2beta2PodsMetricSource
-from .models.v2beta2_pods_metric_status import V2beta2PodsMetricStatus
-from .models.v2beta2_resource_metric_source import V2beta2ResourceMetricSource
-from .models.v2beta2_resource_metric_status import V2beta2ResourceMetricStatus
-from .models.version_info import VersionInfo
+__version__ = "9.0.0-snapshot"
```
Comment (scottilee) [inline kubernetes/client/__init__.py]: Shoudn't this be: "10.0.0-snapshot"?

File: kubernetes/client/api/__init__.py (hunk 2 above)
Comment (scottilee) [reply kubernetes/client/api/__init__.py]: Out of curiosity, why did you change it from apis to api?

File: kubernetes/README.md [hunk 12]
```diff
@@ -68,7 +70,7 @@ except ApiException as e:
 
 ## Documentation for API Endpoints
 
-All URIs are relative to *https://localhost*
+All URIs are relative to *http://localhost*
```
Comment (scottilee) [inline kubernetes/README.md]: Also curious why it's http now instead of https?

**Comment by scottilee** [comment]
The `kubernetes/README.md` Requirements section says "Python 2.7 and 3.4+" but 3.4 is deprecated (refer to https://github.com/kubernetes-client/python/pull/807). Is there a way to fix this?

File: kubernetes/.openapi-generator/COMMIT (hunk 9 above)
Comment (micw523) [reply kubernetes/.openapi-generator/COMMIT]: I believe that we rarely bump the version of swagger-codegen right now.

File: kubernetes/client/api/authentication_v1_api.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (roycaihw) [reply kubernetes/client/api/authentication_v1_api.py]: This was just removing the header for those requests that don't come with a body https://github.com/kubernetes-client/python/pull/931#discussion_r322935538

**Comment by roycaihw** [comment]
we've migrated to openapi-generator

/close
